### PR TITLE
Support for the polygon basic shape in offsetPath. Only supports pixels.

### PR DESCRIPTION
**Pull request by courtneycb**
No support for fill-rule or percentages. 
I am also suspicious of the last test in test/basicShapePolygonTest.js. At 0.5 I was expecting the transform to be 'translate3d(250px, 400px, 0px) rotate(90deg)'. This is because I have purposely made the polygon to be a hexagon with sides of length 250px. However when I console.log the path length it returns 1800, not 1500.

### CODE REVIEW

File: test/testFunctions.js [hunk 1]
```diff
@@ -86,7 +86,6 @@
       animation.currentTime = at;
       var result = target.style._getAnimated(propertyToRead);
       animation.cancel();
-
       assert.equal(result, is, 'For: ' + JSON.stringify(keyframes) + ' at: ' + at + '\n');
```
Comment (ewilligers) [inline test/testFunctions.js]: You should put the blank line back as there isn't a good reason for this patch to change this file.

File: src/offsetPath.js [hunk 2]
```diff
@@ -2,30 +2,32 @@
...
+      coordinate = argumentList[i].trim().split(/\s+/);
+      if (coordinate.length !== 2) {
+        return undefined;
+      }
+      x = internalScope.offsetDistanceParse(coordinate[0]);
+      y = internalScope.offsetDistanceParse(coordinate[1]);
+      if (!x || !y || x.unit === '%' || y.unit === '%') {
+        return undefined;
+      }
+      path += ' h ' + (x.value - previousX) + ' v ' + (y.value - previousY);
```
Comment (ewilligers) [inline src/offsetPath.js]: path += ' l ' + (x.value - previousX) + ' ' + (y.value - previousY);

h 3 v 4 moves horizontally and then vertically, giving a path length of 7.

l 3 4 moves diagonally, giving a path length of 5

**Comment by courtneycb** [comment]
I don't understand the rotations for the last test. Shouldn't it be 150deg at 1 / 3 and 300deg at 2 / 3?

File: src/offsetPath.js [hunk 3]
```diff
@@ -2,6 +2,34 @@
 'use strict';
 
 (function () {
+  function basicShapePolygon (input) {
+    // TODO: Support the fill-rule option and %
```
Comment (ewilligers) [inline src/offsetPath.js]: By the way, fill-rule will have no effect. If it is present, we will check it is valid, and store it so we can serialize, but it won't affect the transform.  (No action needed for this pull request.)

File: test/basicShapePolygonTest.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                                    {at: 0, is: 'translate3d(0px, 0px, 0px)'},
+                                    {at: 0.25, is: 'translate3d(50px, 0px, 0px) rotate(-90deg)'},
+                                    {at: 0.5, is: 'translate3d(50px, -50px, 0px) rotate(180deg)'},
+                                    {at: 0.75, is: 'translate3d(0px, -50px, 0px) rotate(90deg)'},
+                                    {at: 1, is: 'translate3d(0px, 0px, 0px)'}
+        ]
+      );
+
+      assertTransformInterpolation([
+                                    {'offsetPath': 'polygon(250px 0px, 400px 200px, 250px 400px, 0px 400px, -150px 200px)', 'offsetDistance': '0%'},
```
Comment (ewilligers) [inline test/basicShapePolygonTest.js]: Perhaps add an offsetPathParse test for this polygon.

File: src/offsetPath.js [hunk 5]
```diff
@@ -2,6 +2,34 @@
...
+    var argumentList = input.split(',');
+    var coordinate = null;
+    var x = null;
+    var y = null;
+    var previousX = 0;
+    var previousY = 0;
+    var path = 'm 0 0';
```
Comment (ewilligers) [inline src/offsetPath.js]: I suspect we should start with empty string and have m x y instead of l x y for the first point.

File: test/basicShapePolygonTest.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
+/* global suite test assert internalScope */
+'use strict';
+
+(function () {
+  suite('offsetPath', function () {
+    test('polygon', function () {
+      var offsetPathParse = internalScope.offsetPathParse;
+      var assertTransformInterpolation = internalScope.assertTransformInterpolation;
+
+      var result = offsetPathParse('polygon(150px 200px, 250px 150px, 200px -150px)');
```
Comment (ewilligers) [inline test/basicShapePolygonTest.js]: I suspect when we make my m x y change, you will need to start with polygon(0px 0px, ...) to get the current parse result and transform results.

File: test/basicShapePolygonTest.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        ]
+      );
+
+      assertTransformInterpolation([
+                                    {'offsetPath': 'polygon(250px 0px, 400px 200px, 250px 400px, 0px 400px, -150px 200px)', 'offsetDistance': '0%'},
+                                    {'offsetPath': 'polygon(250px 0px, 400px 200px, 250px 400px, 0px 400px, -150px 200px)', 'offsetDistance': '100%'}],
+        [
+                                    {at: 0, is: 'translate3d(0px, 0px, 0px)'},
+                                    {at: 1 / 3, is: 'translate3d(400px, 200px, 0px) rotate(127.04deg)'},
+                                    {at: 0.5, is: 'translate3d(250px, 400px, 0px) rotate(180deg)'},
```
Comment (ewilligers) [inline test/basicShapePolygonTest.js]: I think these rotations are fine. You would only have angles like 150deg if the coordinates involved sqrt(3). If you want simple angles and simple coordinates, your only options might be: isosceles right triangles, rectangles, octagons.

File: src/offsetPath.js [hunk 8]
```diff
@@ -50,6 +83,16 @@
     } else if (path !== null) {
       var pathInput = path[1];
       return {type: 'path', path: pathInput};
+    } else {
+      var basicShapes = ['inset', 'circle', 'ellipse', 'polygon'];
+      if (!isInArray(basicShapes, shapeType[0])) {
```
Comment (alancutter) [inline src/offsetPath.js]: I think shapeType can be null causing this to throw. There should be a test case for this condition.

File: src/offsetPath.js [hunk 9]
```diff
@@ -50,6 +83,16 @@
...
+        return undefined;
+      }
+
+      var shapeArguments = /\(([^)]+)\)/.exec(input);
+      if (shapeType[0] === 'polygon') {
+        return basicShapePolygon(shapeArguments[1]);
```
Comment (alancutter) [inline src/offsetPath.js]: shapeArguments can be null here too.

File: src/offsetPath.js [hunk 10]
```diff
@@ -2,6 +2,38 @@
 'use strict';
 
 (function () {
+  function basicShapePolygon (input) {
```
Comment (alancutter) [inline src/offsetPath.js]: The word "parse" ought to appear somewhere in the function name.

File: src/offsetPath.js [hunk 11]
```diff
@@ -13,9 +45,10 @@
 
     var ray = /^ray\((.*)\)$/.exec(input);
```
Comment (alancutter) [inline src/offsetPath.js]: Code clean up idea:

    function offsetPathParse(input) {
      for (parse of [parseNone, parseRay, parsePath, parseCircle...]) {
        var result = parse(input);
        if (result) {
          return result;
        }
      }
    }

File: src/offsetPath.js (hunk 8 above)
Comment (courtneycb) [inline src/offsetPath.js]: Covered in new patch.

File: src/offsetPath.js (hunk 9 above)
Comment (courtneycb) [inline src/offsetPath.js]: Covered in new patch.

File: src/offsetPath.js (hunk 10 above)
Comment (courtneycb) [inline src/offsetPath.js]: Covered in new patch.

File: src/offsetPath.js (hunk 11 above)
Comment (courtneycb) [inline src/offsetPath.js]: Will do this in a separate patch.
